src/wait_list.rs

```diff
@@ -0,0 +1,474 @@
+        let head = unsafe { NonNull::from(head).as_ref() };
+
+        // Dequeue the first waiter now that it's not necessary to keep it in the queue.
+        unsafe { self.dequeue(head) };
+
+        // Wake the waker last, to ensure that if this panics nothing goes wrong.
+        waker.wake();
+    }
+}
+
+impl<T> Drop for Borrowed<'_, T> {
+    fn drop(&mut self) {
+        debug_assert!(self.list.borrowed.get());
+        self.list.borrowed.set(false);
+    }
+}
+
+/// The inner future used by a waiting operation.
+struct WaitInner<'list, 'waiter, T> {
+    list: &'list WaitList<T>,
+    waiter: &'waiter UnsafeCell<Waiter<T>>,
+}
+
+impl<'list, 'waiter, T> WaitInner<'list, 'waiter, T> {
+    /// # Safety
+    ///
+    /// - `waiter` must be the only reference to that object.
+    unsafe fn new(list: &'list WaitList<T>, waiter: &'waiter UnsafeCell<Waiter<T>>) -> Self {
+        // SAFETY: Upheld by the caller.
+        unsafe { list.borrow().enqueue(waiter) };
+        Self { list, waiter }
+    }
+}
+
+impl<T> Future for WaitInner<'_, '_, T> {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let _guard = self.list.borrow();
+
+        let old_waker = &mut unsafe { &mut *self.waiter.get() }.waker;
+        match old_waker {
+            // No need to update the waker
+            Some(same_waker) if same_waker.will_wake(cx.waker()) => {}
+
+            // Replace the old waker with the current one
+            Some(_) => *old_waker = Some(cx.waker().clone()),
+
+            // No waker means we have been dequeued
+            None => return Poll::Ready(()),
+        }
+        Poll::Pending
+    }
+}
+
+impl<T> Drop for WaitInner<'_, '_, T> {
+    fn drop(&mut self) {
+        let mut list = match self.list.try_borrow() {
+            Some(guard) => guard,
+            // Panicking isn't enough because that would allow the `waiter` to be used after it's
+            // freed.
+            None => process::abort(),
+        };
+
+        if unsafe { &*self.waiter.get() }.waker.is_some() {
+            unsafe { list.dequeue(self.waiter) };
+        }
+    }
+}
+
+struct CloneWaker;
+impl Future for CloneWaker {
+    type Output = task::Waker;
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(cx.waker().clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::future::Future;
+    use std::sync::Arc;
+    use std::task;
+    use std::task::Poll;
+
+    use super::WaitList;
+
+    #[test]
+    fn wake_empty() {
+        let list = <WaitList<()>>::new();
+        list.borrow().wake_one();
+        list.borrow().wake_one();
+        assert_eq!(list.borrow().head_state(), None);
+    }
+
+    #[test]
+    fn cancel() {
+        noop_cx!(cx);
+
+        let list = <WaitList<Box<u32>>>::new();
+        let mut future = Box::pin(list.wait(Box::new(5)));
```

Comment:
I can't think of one which would matter, but is there any reason you're preferring to pin on the heap rather than on the stack for these tests?

Reply:
The real reason is that I didn't realize Tokio was a dependency 😅. I would be happy to change it though.

Reply:
Not important, just curious!